2017/proposals/chait-kaitlyn.md

```diff
@@ -0,0 +1,306 @@
+*Model-View-Controller* pattern as described before.
+
+    *Controller*: In the following prototypes below, an object called
+    Controller being used.  The design purpose of this object is to serve as
+    a data controller. The two following actions are mandatory:
+        * import data from a serialized representation
+        * export data to a serialized representation
+    This object must be implemented with a way to properly unpack/pack
+    data from/to a json object file.
+
+Note:
+    * The set of data can be list, ndarray, or other representation compatible
+    with Matplotlib.  When exporting data, my first focus will be *to_json*
+    function, in the future I would like to work out to other data types.
+    * Keep in mind unserializable aspects of a figure: how to deal with it?
+
+In the implementation, the controller will be key, but then we will have to
+deal with the serialization in PythreeJS.
+
+### Prototype:
+Here are some examples of how the serializer should work and be implemented
+in a PythreeJS widget
+
+1. Display Matplotlib figure from serialized representation:
+~~~
+import json
+from matplotlib.controllers import Controller
+with open('my_figure') as f:
+    o = json.load(f)
+c = Controller(o)
+fig = c.figure
+~~~
+
+2. Export Matplotlib figure to json:
+~~~
+o = c.to_json()
+# or... we should be able to throw a figure object in there too
+o = Controller.to_json(mpl_fig)
+~~~
+[src](https://github.com/matplotlib/matplotlib/wiki/MEP25#examples)
+
+3. Render serialized data into a PythreeJS widget:
+~~~
+#Import Statements
+import pythreejs as py3
+from ipywidgets import HTML, Text, Controller
+from traitlets import link, dlink
+from traitlets.traitlets import _validate_link
+
+import json
+from matplotlib.controllers import Controller
+with open('my_figure') as f:
+    o = json.load(f)
+c = Controller(o)
+
+#Let's say for this example we have spherical data
+#We would need a variable texture that would properly
+#'encompass' a sphereical object.
+
+sphereGeometry = py3.SphereGeometry(radius = 1)
+sphereMaterial = py3.BasicMaterial(map = texture)
+sphere = py3.Mesh(geometry = sphereGeometry, material = sphereMaterial)
+
+camera = py3.PerspectiveCamera(position=[2,0,2])
+scene = py3.Scene(children=[])
+orbit_controls = py3.OrbitControls(controlling=c)
+renderer = py3.Renderer(renderer_type='webgl',camera=c, scene= s,
+                            background='black', controls=[orbit_controls])
+return display(renderer)
+~~~
+
+Note: The output of 3 would_be a PythreeJS Widget containing the figure.
+The following image is a sample sphere I made using PythreeJS.
+
+![PythreeJS_Sphere_Image](https://katierose1029.github.io/gsoc/PythreeJS_Sphere_Image.png)
+
+
+________________________________________________________________________________
+
+## Schedule of Deliverables
+
+**Community Bonding Period:** (May 1st - May 28th)
+
+**Phase 1:** (May 29th - June 23rd)
```

Comment:
Kinda - it's more of an agile/TDD approach, so more prototype serialization/iterate. But yes, it'd be more on target for serialization to take at least two phases and then push the rest of the project into "future work"